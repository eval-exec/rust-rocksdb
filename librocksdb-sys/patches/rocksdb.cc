#include "patches/rocksdb.h"

#include "rocksdb/utilities/options_util.h"

using rocksdb::Cache;
using rocksdb::ColumnFamilyDescriptor;
using rocksdb::Env;
using rocksdb::Options;
using rocksdb::Status;

extern "C" {
    // Copy structs from librocksdb-sys/rocksdb/db/c.cc
    struct rocksdb_cache_t {
        std::shared_ptr<Cache>  rep;
    };
    struct rocksdb_env_t {
        Env* rep;
        bool is_default;
    };
    struct rocksdb_options_t {
        Options rep;
    };

    // New structs
    struct rocksdb_column_family_descriptor_t {
        char *name;
        Options options;
    };
    struct rocksdb_column_family_descriptors_t {
        std::vector<rocksdb_column_family_descriptor_t> rep;
    };

    rocksdb_cache_t* rocksdb_null_cache() {
        rocksdb_cache_t* c = new rocksdb_cache_t;
        c->rep = nullptr;
        return c;
    }

    rocksdb_options_t* rocksdb_options_clone(rocksdb_options_t* options) {
        rocksdb_options_t* o = new rocksdb_options_t;
        o->rep = Options(options->rep);
        return o;
    }

    rocksdb_column_family_descriptors_t* rocksdb_column_family_descriptors_create() {
        return new rocksdb_column_family_descriptors_t;
    }

    void rocksdb_column_family_descriptors_destroy(rocksdb_column_family_descriptors_t* cf_descs) {
        int size = static_cast<int>(cf_descs->rep.size());
        for (int i = 0; i < size; i++) {
            free(cf_descs->rep[i].name);
        }
        delete cf_descs;
    }

    int rocksdb_column_family_descriptors_count(const rocksdb_column_family_descriptors_t* cf_descs) {
        return static_cast<int>(cf_descs->rep.size());
    }

    char* rocksdb_column_family_descriptors_name(const rocksdb_column_family_descriptors_t* cf_descs, int index) {
        return cf_descs->rep[index].name;
    }

    rocksdb_options_t* rocksdb_column_family_descriptors_options(const rocksdb_column_family_descriptors_t* cf_descs, int index) {
        rocksdb_options_t* options = new rocksdb_options_t;
        options->rep = cf_descs->rep[index].options;
        return options;
    }

    rocksdb_fulloptions_t rocksdb_options_load_from_file(
        const char* config_file,
        rocksdb_env_t* env,
        bool ignore_unknown_options,
        rocksdb_cache_t* cache,
        char** errptr) {

        rocksdb_fulloptions_t full_opts;
        full_opts.db_opts = nullptr;
        full_opts.cf_descs = nullptr;

        rocksdb_options_t* db_opts = new rocksdb_options_t;
        std::vector<ColumnFamilyDescriptor> cf_descs_tmp;

        rocksdb::ConfigOptions config_opt;
        config_opt.ignore_unknown_options = ignore_unknown_options;
        config_opt.input_strings_escaped = true;
        config_opt.env = env->rep;

        Status status = rocksdb::LoadOptionsFromFile(
            config_opt,
            std::string(config_file),
            &db_opts->rep,
            &cf_descs_tmp,
            &cache->rep);
        if (status.ok()) {
            rocksdb_column_family_descriptors_t* cf_descs = new rocksdb_column_family_descriptors_t;
            full_opts.db_opts = db_opts;
            int cf_descs_tmp_size = static_cast<int>(cf_descs_tmp.size());
            for (int i = 0; i < cf_descs_tmp_size; i++) {
                rocksdb_column_family_descriptor_t cf_desc;
                cf_desc.name = strdup(cf_descs_tmp[i].name.c_str());
                cf_desc.options = Options(db_opts->rep, cf_descs_tmp[i].options);
                cf_descs->rep.push_back(cf_desc);
            }
            full_opts.cf_descs = cf_descs;
            return full_opts;
        } else {
            delete db_opts;
        }
        if (*errptr != nullptr) {
            free(*errptr);
        }
        *errptr = strdup(status.ToString().c_str());
        return full_opts;
    }
}
