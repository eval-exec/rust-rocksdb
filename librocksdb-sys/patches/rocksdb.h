#include <stdbool.h>

#include "../rocksdb/include/rocksdb/c.h"

#ifdef __cplusplus
extern "C" {
#endif

typedef struct rocksdb_column_family_descriptor_t   rocksdb_column_family_descriptor_t;
typedef struct rocksdb_column_family_descriptors_t  rocksdb_column_family_descriptors_t;
typedef struct {
    rocksdb_options_t* db_opts;
    rocksdb_column_family_descriptors_t* cf_descs;
} rocksdb_fulloptions_t;

extern ROCKSDB_LIBRARY_API
    rocksdb_cache_t* rocksdb_null_cache();

extern ROCKSDB_LIBRARY_API
    rocksdb_options_t* rocksdb_options_clone(rocksdb_options_t* options);

extern ROCKSDB_LIBRARY_API
    rocksdb_column_family_descriptors_t* rocksdb_column_family_descriptors_create();

extern ROCKSDB_LIBRARY_API
    void rocksdb_column_family_descriptors_destroy(rocksdb_column_family_descriptors_t* cf_descs);

extern ROCKSDB_LIBRARY_API
    int rocksdb_column_family_descriptors_count(const rocksdb_column_family_descriptors_t* cf_descs);

extern ROCKSDB_LIBRARY_API
    char* rocksdb_column_family_descriptors_name(const rocksdb_column_family_descriptors_t* cf_descs, int index);

extern ROCKSDB_LIBRARY_API
    rocksdb_options_t* rocksdb_column_family_descriptors_options(const rocksdb_column_family_descriptors_t* cf_descs, int index);

extern ROCKSDB_LIBRARY_API
    rocksdb_fulloptions_t rocksdb_options_load_from_file(
        const char* config_file,
        rocksdb_env_t* env,
        bool ignore_unknown_options,
        rocksdb_cache_t* cache,
        char** errptr);

#ifdef __cplusplus
}  /* end extern "C" */
#endif
