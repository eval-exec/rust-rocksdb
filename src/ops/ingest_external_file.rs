use crate::ffi;
use crate::ffi_util::to_cpath;
use crate::{handle::Handle, ColumnFamily, Error, IngestExternalFileOptions};
use std::ffi::CString;
use std::path::Path;

pub trait IngestExternalFile {
    fn ingest_external_file_full<P: AsRef<Path>>(
        &self,
        paths: Vec<P>,
        opts: Option<&IngestExternalFileOptions>,
    ) -> Result<(), Error>;

    /// Loads a list of external SST files created with SstFileWriter into the DB with default opts
    fn ingest_external_file<P: AsRef<Path>>(&self, paths: Vec<P>) -> Result<(), Error> {
        self.ingest_external_file_full(paths, None)
    }

    /// Loads a list of external SST files created with SstFileWriter into the DB
    fn ingest_external_file_opts<P: AsRef<Path>>(
        &self,
        paths: Vec<P>,
        opts: &IngestExternalFileOptions,
    ) -> Result<(), Error> {
        self.ingest_external_file_full(paths, Some(opts))
    }
}

pub trait IngestExternalFileCF {
    fn ingest_external_file_cf_full<P: AsRef<Path>>(
        &self,
        cf: Option<&ColumnFamily>,
        paths: Vec<P>,
        opts: Option<&IngestExternalFileOptions>,
    ) -> Result<(), Error>;

    /// Loads a list of external SST files created with SstFileWriter into the DB for given Column Family
    /// with default opts
    fn ingest_external_file_cf<P: AsRef<Path>>(
        &self,
        cf: &ColumnFamily,
        paths: Vec<P>,
    ) -> Result<(), Error> {
        self.ingest_external_file_cf_full(Some(cf), paths, None)
    }

    /// Loads a list of external SST files created with SstFileWriter into the DB for given Column Family
    fn ingest_external_file_opts<P: AsRef<Path>>(
        &self,
        cf: &ColumnFamily,
        paths: Vec<P>,
        opts: &IngestExternalFileOptions,
    ) -> Result<(), Error> {
        self.ingest_external_file_cf_full(Some(cf), paths, Some(opts))
    }
}

impl<T> IngestExternalFile for T
where
    T: IngestExternalFileCF,
{
    fn ingest_external_file_full<P: AsRef<Path>>(
        &self,
        paths: Vec<P>,
        opts: Option<&IngestExternalFileOptions>,
    ) -> Result<(), Error> {
        self.ingest_external_file_cf_full(None, paths, opts)
    }
}

impl<T> IngestExternalFileCF for T
where
    T: Handle<ffi::rocksdb_t> + super::Write,
{
    fn ingest_external_file_cf_full<P: AsRef<Path>>(
        &self,
        cf: Option<&ColumnFamily>,
        paths: Vec<P>,
        opts: Option<&IngestExternalFileOptions>,
    ) -> Result<(), Error> {
        let paths_v: Vec<CString> = paths
            .iter()
            .map(|path| {
                to_cpath(
                    &path,
                    "Failed to convert path to CString when IngestExternalFile.",
                )
            })
            .collect::<Result<Vec<_>, _>>()?;

        let cpaths: Vec<_> = paths_v.iter().map(|path| path.as_ptr()).collect();

        let mut default_opts = None;

        let ief_handle = IngestExternalFileOptions::input_or_default(opts, &mut default_opts)?;

        unsafe {
            match cf {
                Some(cf) => ffi_try!(ffi::rocksdb_ingest_external_file_cf(
                    self.handle(),
                    cf.handle(),
                    cpaths.as_ptr(),
                    paths_v.len(),
                    ief_handle
                )),
                None => ffi_try!(ffi::rocksdb_ingest_external_file(
                    self.handle(),
                    cpaths.as_ptr(),
                    paths_v.len(),
                    ief_handle
                )),
            };

            Ok(())
        }
    }
}
