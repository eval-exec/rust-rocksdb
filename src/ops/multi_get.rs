// Copyright 2019 Tyler Neely
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

use crate::{ffi, ColumnFamily, DBVector};
use libc::c_char;
use std::ptr;

use crate::{handle::Handle, Error, ReadOptions};

pub type CFAndKey<'a> = (&'a ColumnFamily, Box<[u8]>);

pub trait MultiGet<R> {
    fn multi_get_full<K, I>(
        &self,
        keys: I,
        readopts: Option<&R>,
    ) -> Vec<Result<Option<DBVector>, Error>>
    where
        K: AsRef<[u8]>,
        I: IntoIterator<Item = K>;

    fn multi_get<K, I>(&self, keys: I) -> Vec<Result<Option<DBVector>, Error>>
    where
        K: AsRef<[u8]>,
        I: IntoIterator<Item = K>,
    {
        self.multi_get_full(keys, None)
    }

    fn multi_get_opt<K, I>(&self, keys: I, readopts: &R) -> Vec<Result<Option<DBVector>, Error>>
    where
        K: AsRef<[u8]>,
        I: IntoIterator<Item = K>,
    {
        self.multi_get_full(keys, Some(readopts))
    }
}

pub trait MultiGetCF<R> {
    fn multi_get_cf_full<'a, K, I>(
        &self,
        keys_cf: I,
        readopts: Option<&R>,
    ) -> Vec<Result<Option<DBVector>, Error>>
    where
        K: AsRef<[u8]>,
        I: IntoIterator<Item = (&'a ColumnFamily, K)>;

    fn multi_get_cf<'a, K, I>(&self, keys_cf: I) -> Vec<Result<Option<DBVector>, Error>>
    where
        K: AsRef<[u8]>,
        I: IntoIterator<Item = (&'a ColumnFamily, K)>,
    {
        self.multi_get_cf_full(keys_cf, None)
    }

    fn multi_get_cf_opt<'a, K, I>(
        &self,
        keys_cf: I,
        readopts: &R,
    ) -> Vec<Result<Option<DBVector>, Error>>
    where
        K: AsRef<[u8]>,
        I: IntoIterator<Item = (&'a ColumnFamily, K)>,
    {
        self.multi_get_cf_full(keys_cf, Some(readopts))
    }
}

impl<T> MultiGet<ReadOptions> for T
where
    T: Handle<ffi::rocksdb_t> + super::Read,
{
    /// Return the values associated with the given keys using read options.
    fn multi_get_full<K, I>(
        &self,
        keys: I,
        readopts: Option<&ReadOptions>,
    ) -> Vec<Result<Option<DBVector>, Error>>
    where
        K: AsRef<[u8]>,
        I: IntoIterator<Item = K>,
    {
        let mut default_readopts = None;
        let ro_handle = match ReadOptions::input_or_default(readopts, &mut default_readopts) {
            Ok(ro) => ro,
            Err(e) => {
                let key_count = keys.into_iter().count();

                return vec![e; key_count]
                    .iter()
                    .map(|e| Err(e.to_owned()))
                    .collect();
            }
        };

        let (keys, keys_sizes): (Vec<Box<[u8]>>, Vec<_>) = keys
            .into_iter()
            .map(|k| (Box::from(k.as_ref()), k.as_ref().len()))
            .unzip();
        let ptr_keys: Vec<_> = keys.iter().map(|k| k.as_ptr() as *const c_char).collect();
        let mut values = vec![ptr::null_mut(); keys.len()];
        let mut values_sizes = vec![0_usize; keys.len()];
        let mut errors = vec![ptr::null_mut(); keys.len()];
        unsafe {
            ffi::rocksdb_multi_get(
                self.handle(),
                ro_handle,
                ptr_keys.len(),
                ptr_keys.as_ptr(),
                keys_sizes.as_ptr(),
                values.as_mut_ptr(),
                values_sizes.as_mut_ptr(),
                errors.as_mut_ptr(),
            );
        }
        convert_values(values, values_sizes, errors)
    }
}

impl<T> MultiGetCF<ReadOptions> for T
where
    T: Handle<ffi::rocksdb_t> + super::Read,
{
    /// Return the values associated with the given keys and column families using read options.
    fn multi_get_cf_full<'a, K, I>(
        &self,
        keys: I,
        readopts: Option<&ReadOptions>,
    ) -> Vec<Result<Option<DBVector>, Error>>
    where
        K: AsRef<[u8]>,
        I: IntoIterator<Item = (&'a ColumnFamily, K)>,
    {
        let mut default_readopts = None;
        let ro_handle = match ReadOptions::input_or_default(readopts, &mut default_readopts) {
            Ok(ro) => ro,
            Err(e) => {
                let key_count = keys.into_iter().count();
                return vec![e; key_count]
                    .iter()
                    .map(|e| Err(e.to_owned()))
                    .collect();
            }
        };

        let (cfs_and_keys, keys_sizes): (Vec<CFAndKey>, Vec<_>) = keys
            .into_iter()
            .map(|(cf, key)| ((cf, Box::from(key.as_ref())), key.as_ref().len()))
            .unzip();
        let ptr_keys: Vec<_> = cfs_and_keys
            .iter()
            .map(|(_, k)| k.as_ptr() as *const c_char)
            .collect();
        let ptr_cfs: Vec<_> = cfs_and_keys
            .iter()
            .map(|(c, _)| c.inner as *const _)
            .collect();

        let mut values = vec![ptr::null_mut(); ptr_keys.len()];
        let mut values_sizes = vec![0_usize; ptr_keys.len()];
        let mut errors = vec![ptr::null_mut(); ptr_keys.len()];
        unsafe {
            ffi::rocksdb_multi_get_cf(
                self.handle(),
                ro_handle,
                ptr_cfs.as_ptr(),
                ptr_keys.len(),
                ptr_keys.as_ptr(),
                keys_sizes.as_ptr(),
                values.as_mut_ptr(),
                values_sizes.as_mut_ptr(),
                errors.as_mut_ptr(),
            );
        }

        convert_values(values, values_sizes, errors)
    }
}

pub fn convert_values(
    values: Vec<*mut c_char>,
    values_sizes: Vec<usize>,
    errors: Vec<*mut c_char>,
) -> Vec<Result<Option<DBVector>, Error>> {
    values
        .into_iter()
        .zip(values_sizes.into_iter())
        .zip(errors.into_iter())
        .map(|((v, s), e)| {
            if e.is_null() {
                if v.is_null() {
                    return Ok(None);
                }
                unsafe { Ok(Some(DBVector::from_c(v as *mut u8, s))) }
            } else {
                Err(Error::new(crate::ffi_util::error_message(e)))
            }
        })
        .collect()
}
