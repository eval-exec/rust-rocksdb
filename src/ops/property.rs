use crate::{handle::Handle, ColumnFamily, Error};
use libc::{c_char, c_void};
use std::ffi::{CStr, CString};

pub trait GetProperty {
    /// Retrieves a RocksDB property by name.
    ///
    /// For a full list of properties, see
    /// https://github.com/facebook/rocksdb/blob/08809f5e6cd9cc4bc3958dd4d59457ae78c76660/include/rocksdb/db.h#L428-L634
    fn property_value(&self, name: &str) -> Result<Option<String>, Error>;

    /// Retrieves a RocksDB property and casts it to an integer.
    ///
    /// For a full list of properties that return int values, see
    /// https://github.com/facebook/rocksdb/blob/08809f5e6cd9cc4bc3958dd4d59457ae78c76660/include/rocksdb/db.h#L654-L689
    fn property_int_value(&self, name: &str) -> Result<Option<u64>, Error>;
}

pub trait GetPropertyCF {
    /// Retrieves a RocksDB property by name, for a specific column family.
    ///
    /// For a full list of properties, see
    /// https://github.com/facebook/rocksdb/blob/08809f5e6cd9cc4bc3958dd4d59457ae78c76660/include/rocksdb/db.h#L428-L634
    fn property_value_cf(&self, cf: &ColumnFamily, name: &str) -> Result<Option<String>, Error>;

    /// Retrieves a RocksDB property for a specific column family and casts it to an integer.
    ///
    /// For a full list of properties that return int values, see
    /// https://github.com/facebook/rocksdb/blob/08809f5e6cd9cc4bc3958dd4d59457ae78c76660/include/rocksdb/db.h#L654-L689
    fn property_int_value_cf(&self, cf: &ColumnFamily, name: &str) -> Result<Option<u64>, Error>;
}

impl<T> GetProperty for T
where
    T: Handle<ffi::rocksdb_t>,
{
    fn property_value(&self, name: &str) -> Result<Option<String>, Error> {
        property_value_impl(
            name,
            |prop_name| unsafe { ffi::rocksdb_property_value(self.handle(), prop_name) },
            |str_value| Ok(str_value.to_owned()),
        )
    }

    fn property_int_value(&self, name: &str) -> Result<Option<u64>, Error> {
        property_value_impl(
            name,
            |prop_name| unsafe { ffi::rocksdb_property_value(self.handle(), prop_name) },
            parse_property_int_value,
        )
    }
}

impl<T> GetPropertyCF for T
where
    T: Handle<ffi::rocksdb_t>,
{
    fn property_value_cf(&self, cf: &ColumnFamily, name: &str) -> Result<Option<String>, Error> {
        property_value_impl(
            name,
            |prop_name| unsafe {
                ffi::rocksdb_property_value_cf(self.handle(), cf.inner, prop_name)
            },
            |str_value| Ok(str_value.to_owned()),
        )
    }

    fn property_int_value_cf(&self, cf: &ColumnFamily, name: &str) -> Result<Option<u64>, Error> {
        property_value_impl(
            name,
            |prop_name| unsafe {
                ffi::rocksdb_property_value_cf(self.handle(), cf.inner, prop_name)
            },
            parse_property_int_value,
        )
    }
}

fn parse_property_int_value(value: &str) -> Result<u64, Error> {
    value.parse::<u64>().map_err(|err| {
        Error::new(format!(
            "Failed to convert property value {} to int: {}",
            value, err
        ))
    })
}

/// Implementation for property_value et al methods.
///
/// `name` is the name of the property.  It will be converted into a CString
/// and passed to `get_property` as argument.  `get_property` reads the
/// specified property and either returns NULL or a pointer to a C allocated
/// string; this method takes ownership of that string and will free it at
/// the end.  That string is parsed using `parse` callback which produces
/// the returned result.
fn property_value_impl<R>(
    name: &str,
    get_property: impl FnOnce(*const c_char) -> *mut c_char,
    parse: impl FnOnce(&str) -> Result<R, Error>,
) -> Result<Option<R>, Error> {
    let value = match CString::new(name) {
        Ok(prop_name) => get_property(prop_name.as_ptr()),
        Err(e) => {
            return Err(Error::new(format!(
                "Failed to convert property name to CString: {}",
                e
            )));
        }
    };
    if value.is_null() {
        return Ok(None);
    }
    let result = match unsafe { CStr::from_ptr(value) }.to_str() {
        Ok(s) => parse(s).map(Some),
        Err(e) => Err(Error::new(format!(
            "Failed to convert property value to string: {}",
            e
        ))),
    };
    unsafe {
        ffi::rocksdb_free(value as *mut c_void);
    }
    result
}
