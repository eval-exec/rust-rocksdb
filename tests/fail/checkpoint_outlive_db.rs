extern crate ckb_rocksdb as rocksdb;

use rocksdb::{DB, checkpoint::Checkpoint};
use rocksdb::ops::Open;

fn main() {
    let _checkpoint = {
        let db = DB::open_default("foo").unwrap();
        Checkpoint::new(&db)
    };
}
