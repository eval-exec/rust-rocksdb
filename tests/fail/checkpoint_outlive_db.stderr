error[E0597]: `db` does not live long enough
  --> $DIR/checkpoint_outlive_db.rs:9:25
   |
7  |     let _checkpoint = {
   |         ----------- borrow later stored here
8  |         let db = DB::open_default("foo").unwrap();
9  |         Checkpoint::new(&db)
   |                         ^^^ borrowed value does not live long enough
10 |     };
   |     - `db` dropped here while still borrowed
